Sources/NIOHTTP2/HTTP2ChannelHandler.swift

```diff
@@ -699,8 +807,51 @@ NIOHTTP2Handler(
     mode: \(String(describing: self.mode)),
     initialSettings: \(String(describing: self.initialSettings)),
     channelClosed: \(String(describing: self.channelClosed)),
-    channelWritable: \(String(describing: self.channelWritable))
+    channelWritable: \(String(describing: self.channelWritable)),
+    activationState: \(String(describing: self.activationState))
 )
 """
     }
 }
+
+extension NIOHTTP2Handler {
+    /// Tracks the state of activation of the handler.
+    enum ActivationState {
+        /// The handler hasn't been activated yet.
+        case idle
+
+        /// The handler has received channelActive, but hasn't yet fired it on.
+        case activating
+
+        /// The handler was activating when it received channelInactive. The channel
+        /// must go inactive after firing channel active.
+        case inactiveWhileActivating
+
+        /// The channel has received and fired channelActive.
+        case activated
+
+        /// The channel has received and fired channelActive and channelInactive.
+        case inactive
+    }
+
+    fileprivate func impossibleActivationStateTransition(
+        state: ActivationState, activating: Bool, context: ChannelHandlerContext
+    ) {
+        debugOnly {
+            if !self.tolerateImpossibleStateTransitionsInDebugMode {
+                fatalError("Unexpected channelActive in state \(state)")
+            }
+        }
```

Comment:
Isn't this just `assert(self.tolerateImpossibleStateTransitionsInDebugMode, "Unexpected channelActive in state \(state)")`?

Reply:
Yes, it is. I had done this more awkward spelling because _originally_ the code would always fatalError, so I had nothing to assert on. When I realised some of these code paths could actually be tested I refactored and didn't change the spelling.